### PR TITLE
[SPARK-12049] [CORE] User JVM shutdown hook can cause deadlock at shutdown

### PR DESCRIPTION
**Pull request by srowen**
Avoid potential deadlock with a user app's shutdown hook thread by more narrowly synchronizing access to 'hooks'

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala [hunk 1]
```diff
@@ -232,22 +232,23 @@ private [util] class SparkShutdownHookManager {
     }
   }
 
-  def runAll(): Unit = synchronized {
+  def runAll(): Unit = {
     shuttingDown = true
-    while (!hooks.isEmpty()) {
-      Try(Utils.logUncaughtExceptions(hooks.poll().run()))
+    var nextHook: SparkShutdownHook = null
+    while ({nextHook = hooks synchronized { hooks.poll() }; nextHook != null}) {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala]: nit: we generally use `hooks.synchronized`.

File: core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala [hunk 2]
```diff
@@ -232,22 +232,23 @@ private [util] class SparkShutdownHookManager {
...
+      Try(Utils.logUncaughtExceptions(nextHook.run()))
     }
   }
 
-  def add(priority: Int, hook: () => Unit): AnyRef = synchronized {
+  def add(priority: Int, hook: () => Unit): AnyRef = {
     checkState()
     val hookRef = new SparkShutdownHook(priority, hook)
-    hooks.add(hookRef)
+    hooks synchronized { hooks.add(hookRef) }
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala]: So, technically, this is not really thread-safe anymore. You can have this order of things:

T1: in add(), passes `checkState()` check
T2: in runAll(), sets `shuttingDown` to true
T2: in runAll(), calls `hooks.poll()`, list is empty, stops calling hooks
T1: in add(), adds hook to the list, hook will never be called.

Granted, that's very unlikely. I think if you move the `checkState()` call inside the lock it should fix things, though. You may end up adding a new hook when `shuttingDown` is true in that case, but you make sure that `runAll` will still pick it up.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala]: Yeah, you got me. That really has to be tightened up. Will address these shortly.

**Comment by vanzin** [comment]
LGTM.

**Comment by vanzin** [comment]
Merging (master, 1.6, 1.5).

**Pull request closed without merging**